vector/src/main/java/im/vector/app/features/widgets/webview/WebviewPermissionUtils.kt

```diff
@@ -37,16 +50,50 @@ object WebviewPermissionUtils {
                     allowedPermissions[which] = allowedPermissions[which].first to isChecked
                 }
                 .setPositiveButton(R.string.room_widget_resource_grant_permission) { _, _ ->
-                    request.grant(allowedPermissions.mapNotNull { perm ->
+                    permissionRequest = request
+                    selectedPermissions = allowedPermissions.mapNotNull { perm ->
                         perm.first.takeIf { perm.second }
-                    }.toTypedArray())
+                    }
+
+                    val requiredAndroidPermissions = selectedPermissions.mapNotNull { permission ->
+                        webPermissionToAndroidPermission(permission)
+                    }
+
+                    // When checkPermissions returns false, some of the required Android permissions will
+                    // have to be requested and the flow completes asynchronously via onPermissionResult
+                    if (checkPermissions(requiredAndroidPermissions, activity, activityResultLauncher)) {
+                        request.grant(selectedPermissions.toTypedArray())
+                        reset()
+                    }
                 }
                 .setNegativeButton(R.string.room_widget_resource_decline_permission) { _, _ ->
                     request.deny()
                 }
                 .show()
     }
 
+    fun onPermissionResult(result: Map<String, Boolean>) {
+        permissionRequest?.let { request ->
```

Comment:
is there a case where we would call `onPermissionResult` without having a valid `permissionRequest` available?

I would lean towards throwing instead of silently swallowing the missing `promptForPermissions` (if the ordering is implicitly required)

Reply:
No, that shouldn't happen, I believe. The order is required as you say. I've added an NPE instead of the `let`. Let me know if that's what you had in mind.